### ISSUE TITLE
Client error: `POST https://appleid.apple.com/auth/token` resulted in a `400 Bad Request` response: {"error":"invalid_client"}

### ISSUE THREAD

**Issue by dhivya-picco** [mikebronner/laravel-sign-in-with-apple#38]
I have configured all the configuration for apple signin. It is redirect to apple site to get username and password. after that in callback, it is showing like this

---
## Acceptance Criteria

- [ ] `invalid_client` error from Apple token endpoint is caught and wrapped in a descriptive exception
- [ ] Package documentation explains how to correctly generate the `client_secret` JWT
- [ ] Optionally: provide a helper command or utility to generate/validate the `client_secret`

### Test Coverage
- [ ] Unit test: `invalid_client` Apple response is caught and rethrown as a descriptive exception
- [ ] Integration test: correct `client_secret` generation flow does not trigger the error

---

## WSJF Priority Score
- **Score:** 3.25
- **User/Business Value:** 5 (Recurring auth failure for misconfigured setups)
- **Time Criticality:** 4 (Steady stream of affected users)
- **Risk Reduction/Opportunity Enablement:** 4 (Better error handling reduces support burden)
- **Job Size:** 4 points (S — error handling + docs)
- **Calculation:** (5 + 4 + 4) ÷ 4 = 3.25

**Comment by haid45** [comment]
I am getting the same issue, did you have any luck resolving this?

**Comment by rohail-office** [comment]
Yes I am getting this same issue before a month its working fine on my site

**Comment by iamaz007** [comment]
It is because, your client_secret token is expired, you need to generate a new token and replace it with old one, it will be fixed then.
you can add max 6 months expire time of JWT token, after 6 months you have to generate new one.
https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens

**Comment by rohail-office** [comment]
@iamaz007 I create new app and new client secret and update the existing configuration but its giving the same error

**Comment by iamaz007** [comment]
> @iamaz007 I create new app and new client secret and update the existing configuration but its giving the same error

you don't need to create new app in apple, just generate a new client_secret, make sure you have selected right algorithm it is "ES256" for apple JWT in code, after implementing new client_secret, clear your Laravel cache, to do this, simply go to boostrap/cache and delete all files from there except .gitIgnore

**Comment by rohail-office** [comment]
Thanks I will try this

**Comment by rohail-office** [comment]
@iamaz007  
Unfortunately I am facing this same issue after removing the Laravel cache  and deleting all files except .gitignore form boostrap/cache folder .

This is how , I am creating my client secret

require 'jwt'

key_file = 'key.txt'
team_id = 'XYZ'
client_id = 'XYZ'
key_id = 'XYZ'

ecdsa_key = OpenSSL::PKey::EC.new IO.read key_file

headers = {
'kid' => key_id
}

claims = {
    'iss' => team_id,
    'iat' => Time.now.to_i,
    'exp' => Time.now.to_i + 86400*180,
    'aud' => 'https://appleid.apple.com',
    'sub' => client_id,
}

token = JWT.encode claims, ecdsa_key, 'ES256', headers
puts token

**Comment by karser** [comment]
I ended up generating the client secret using `lcobucci/jwt`
```
<?php
use Lcobucci\JWT\Configuration;
use Lcobucci\JWT\Signer\Hmac\Sha256;
use Lcobucci\JWT\Signer\Key\InMemory;

$now = new \DateTimeImmutable();

$jwtConfig = Configuration::forSymmetricSigner(
    new Sha256(),
    InMemory::file(__DIR__ . '/AuthKey.pem')
);

$token = $jwtConfig->builder()
    ->issuedBy('XXXXXXXX')
    ->issuedAt($now)
    ->expiresAt($now->modify('+1 hour'))
    ->permittedFor('https://appleid.apple.com')
    ->relatedTo('com.example.service-id')
    ->withHeader('kid', 'XXXXXXXX')
    ->getToken($jwtConfig->signer(), $jwtConfig->signingKey());

echo $token->toString();
```
more info is [here](https://karser.dev/how-to-generate-client-secret-for-apple-sign-in-from-p8-certificate/)

**Comment by KaviiChathuranga** [comment]
> > @iamaz007 I create new app and new client secret and update the existing configuration but its giving the same error
> 
> you don't need to create new app in apple, just generate a new client_secret, make sure you have selected right algorithm it is "ES256" for apple JWT in code, after implementing new client_secret, clear your Laravel cache, to do this, simply go to boostrap/cache and delete all files from there except .gitIgnore

Plz, how to generate new client secret ?

**Comment by origooo** [comment]
> > > @iamaz007 I create new app and new client secret and update the existing configuration but its giving the same error
> > 
> > 
> > you don't need to create new app in apple, just generate a new client_secret, make sure you have selected right algorithm it is "ES256" for apple JWT in code, after implementing new client_secret, clear your Laravel cache, to do this, simply go to boostrap/cache and delete all files from there except .gitIgnore
> 
> Plz, how to generate new client secret ?

I have the following implementation in node, but I guess you can rewrite it pretty easy. It renews the token every second minute, and the token is valid according to www.jwt.io.

`
let appleClientSecretToken
const privateKey = fs.readFileSync(KEY_FILE)

const createAppleClientSecretToken = () => {
  try {
    jwt.verify(appleClientSecretToken, privateKey)
  } catch (error) {
    if (
      error.constructor === JsonWebTokenError ||
      error.constructor === TokenExpiredError
    ) {
      console.info("Renewing/creating Apple client_secret JWT token")

      appleClientSecretToken = jwt.sign(
        {
          iss: TEAM_ID,
          iat: Math.floor(Date.now() / 1000),
          exp: Math.floor(Date.now() / 1000) + 120,
          aud: "https://appleid.apple.com",
          sub: SERVICE_ID,
        },
        privateKey,
        {
          algorithm: "ES256",
          header: {
            alg: "ES256",
            kid: KEY_ID,
          },
        }
      )
    }
  }

  return appleClientSecretToken
}
`

Hope that can help!

**Comment by origooo** [comment]
> Plz, how to generate new client secret ?

I have the following implementation in node, but I guess you can rewrite it pretty easy. It renews the token every second minute, and the token is valid according to www.jwt.io.

```
let appleClientSecretToken
const privateKey = fs.readFileSync(KEY_FILE)

const createAppleClientSecretToken = () => {
  try {
    jwt.verify(appleClientSecretToken, privateKey)
  } catch (error) {
    if (
      error.constructor === JsonWebTokenError ||
      error.constructor === TokenExpiredError
    ) {
      console.info("Renewing/creating Apple client_secret JWT token")

      appleClientSecretToken = jwt.sign(
        {
          iss: TEAM_ID,
          iat: Math.floor(Date.now() / 1000),
          exp: Math.floor(Date.now() / 1000) + 120,
          aud: "https://appleid.apple.com",
          sub: SERVICE_ID,
        },
        privateKey,
        {
          algorithm: "ES256",
          header: {
            alg: "ES256",
            kid: KEY_ID,
          },
        }
      )
    }
  }

  return appleClientSecretToken
}
```

Hope that can help a bit!

**Comment by xeisu** [comment]
This library is so messy. I wasted days to make it work. I have the same issue, regenerated the key and everything but it's about how the return object is handled in this library. Author also doesn't seem to be active anymore, such a shame.

**Comment by mikebronner** [comment]
> This library is so messy. I wasted days to make it work. I have the same issue, regenerated the key and everything but it's about how the return object is handled in this library. Author also doesn't seem to be active anymore, such a shame.

True, there also has been some confusion as to this library being the one used in the socialite providers collection, which it is not. As to the state of this library and my activity, I have not been using it myself in projects since I quit freelancing, so any help in form of PRs is greatly appreciated.

**Comment by davidsalazar** [comment]
regenerating the secret key worked for me. Strange how we have to do this every 6 months.

**Comment by developeralamin** [comment]
> regenerating the secret key worked for me. Strange how apple limits it to 6 months.

Please your full process kindly how to works .  I am getting error .
Client error: `POST https://appleid.apple.com/auth/token` resulted in a `400 Bad Request` response: {"error":"invalid_client","email_verified":false}

**Comment by mikebronner** [comment]
PR #73 approved ✅

Implements descriptive error handling for `invalid_client` and `invalid_grant` responses from Apple's token endpoint, adds a PHP-based `ClientSecretGenerator` utility to replace the Ruby script, and includes comprehensive tests for both. All AC verified against actual code.
